src/typing_extensions.py

```diff
@@ -3599,6 +3602,147 @@ def __eq__(self, other: object) -> bool:
+        want to support earlier Python versions, to simply write:
+
+            typing_extensions.get_annotations(obj, format=Format.FORWARDREF)
+
+        """
+        if format not in (Format.VALUE, Format.FORWARDREF, Format.SOURCE):
+            raise ValueError(f"format must be one of {Format.VALUE}, "
+                             f"{Format.FORWARDREF}, {Format.SOURCE}")
+
+        if eval_str and format != Format.VALUE:
+            raise ValueError("eval_str=True is only supported with format=Format.VALUE")
+
+        if isinstance(obj, type):
+            # class
+            obj_dict = getattr(obj, '__dict__', None)
+            if obj_dict and hasattr(obj_dict, 'get'):
+                ann = obj_dict.get('__annotations__', None)
+                if isinstance(ann, _types.GetSetDescriptorType):
+                    ann = None
+            else:
+                ann = None
+
+            obj_globals = None
+            module_name = getattr(obj, '__module__', None)
+            if module_name:
+                module = sys.modules.get(module_name, None)
+                if module:
+                    obj_globals = getattr(module, '__dict__', None)
+            obj_locals = dict(vars(obj))
+            unwrap = obj
+        elif isinstance(obj, _types.ModuleType):
+            # module
+            ann = getattr(obj, '__annotations__', None)
+            obj_globals = obj.__dict__
+            obj_locals = None
+            unwrap = None
+        elif callable(obj):
+            # this includes types.Function, types.BuiltinFunctionType,
+            # types.BuiltinMethodType, functools.partial, functools.singledispatch,
+            # "class funclike" from Lib/test/test_inspect... on and on it goes.
+            ann = getattr(obj, '__annotations__', None)
+            obj_globals = getattr(obj, '__globals__', None)
+            obj_locals = None
+            unwrap = obj
+        elif hasattr(obj, '__annotations__'):
+            ann = obj.__annotations__
+            obj_globals = obj_locals = unwrap = None
+        else:
+            raise TypeError(f"{obj!r} is not a module, class, or callable.")
+
+        if ann is None:
+            return {}
+
+        if not isinstance(ann, dict):
+            raise ValueError(f"{obj!r}.__annotations__ is neither a dict nor None")
+
+        if not ann:
+            return {}
+
+        if not eval_str:
+            if format == Format.SOURCE:
+                return {
+                    key: value if isinstance(value, str) else typing._type_repr(value)
+                    for key, value in ann.items()
+                }
+            return dict(ann)
+
+        if unwrap is not None:
+            while True:
+                if hasattr(unwrap, '__wrapped__'):
+                    unwrap = unwrap.__wrapped__
+                    continue
+                if isinstance(unwrap, functools.partial):
+                    unwrap = unwrap.func
+                    continue
+                break
```

Comment:
Not sure. Looks like unwrap() doesn't handle functools.partial specially, though unlike this code it deals with cycles.

Reply:
It's pretty odd; no CPython tests fail if I make this change:

```diff
diff --git a/Lib/inspect.py b/Lib/inspect.py
index 2b7f8bec482..4f9c43d9fc7 100644
--- a/Lib/inspect.py
+++ b/Lib/inspect.py
@@ -235,13 +235,11 @@ def get_annotations(obj, *, globals=None, locals=None, eval_str=False):
             if module:
                 obj_globals = getattr(module, '__dict__', None)
         obj_locals = dict(vars(obj))
-        unwrap = obj
     elif isinstance(obj, types.ModuleType):
         # module
         ann = getattr(obj, '__annotations__', None)
         obj_globals = getattr(obj, '__dict__')
         obj_locals = None
-        unwrap = None
     elif callable(obj):
         # this includes types.Function, types.BuiltinFunctionType,
         # types.BuiltinMethodType, functools.partial, functools.singledispatch,
@@ -249,7 +247,6 @@ def get_annotations(obj, *, globals=None, locals=None, eval_str=False):
         ann = getattr(obj, '__annotations__', None)
         obj_globals = getattr(obj, '__globals__', None)
         obj_locals = None
-        unwrap = obj
     else:
         raise TypeError(f"{obj!r} is not a module, class, or callable.")
 
@@ -265,17 +262,8 @@ def get_annotations(obj, *, globals=None, locals=None, eval_str=False):
     if not eval_str:
         return dict(ann)
 
-    if unwrap is not None:
-        while True:
-            if hasattr(unwrap, '__wrapped__'):
-                unwrap = unwrap.__wrapped__
-                continue
-            if isinstance(unwrap, functools.partial):
-                unwrap = unwrap.func
-                continue
-            break
-        if hasattr(unwrap, "__globals__"):
-            obj_globals = unwrap.__globals__
+    if not isinstance(obj, types.ModuleType):
+        obj_globals = getattr(unwrap(obj), "__globals__", None) or obj_globals
```

You could consider that simplification here since we already import `inspect` in `typing_extensions`, though it obviously can't be applied to CPython

Reply:
That sounds like we're missing test coverage for `functools.partial`. I think we should mirror the CPython implementation as much as possible.